commons/src/main/scala/org/bitbucket/inkytonik/cooma/Compiler.scala

```diff
@@ -112,14 +112,14 @@ trait Compiler {
                     letV(a, prmV(argumentP(nArg), Vector()),
                         compileTop(e, nArg + 1))
                 case t =>
-                    def aux(t : Expression) : Seq[String] =
+                    def aux(t : Expression) : List[String] =
                         t match {
                             case Cat(e1, e2) =>
-                                aux(e1) ++ aux(e2)
+                                aux(e1) ::: aux(e2)
```

Comment:
Quick question about the `:::`: doesn't this prepend e2's elements to e1's, so they come out in the reverse order? I suppose it might not matter for type checking, but might be confusing.

Reply:
'`:::`' is right-associative (ends in a colon), so as an infix operator it invokes the method on the right operand with the left operand as the parameter. '`::`' works the same way.

```
scala> val xs0 = List(1, 2, 3)
xs0: List[Int] = List(1, 2, 3)

scala> val xs1 = List(4, 5, 6)
xs1: List[Int] = List(4, 5, 6)

scala> xs0 ::: xs1
res0: List[Int] = List(1, 2, 3, 4, 5, 6)

scala> xs1.:::(xs0)
res1: List[Int] = List(1, 2, 3, 4, 5, 6)

scala> 0 :: xs0
res2: List[Int] = List(0, 1, 2, 3)

scala> xs0.::(0)
res3: List[Int] = List(0, 1, 2, 3)
```

Reply:
Ah yes, too much programming in other languages... I never liked the ":" operator rule in Scala. Is it better than just using ++?

Reply:
Probably leads to fewer virtual table lookups. Functionally, they're identical AFAIK.

Reply:
Yeah, probably. The Scala collection library is nice in many ways, but this sort of complexity makes things more difficult to understand  than I would like.